osu.Game.Rulesets.Osu/OsuDifficulty/Preprocessing/OsuDifficultyBeatmap.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Collections;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Osu.Objects;
+
+namespace osu.Game.Rulesets.Osu.OsuDifficulty.Preprocessing
+{
+    public class OsuDifficultyBeatmap : IEnumerable<OsuDifficultyHitObject>
+    {
+        private readonly IEnumerator<OsuDifficultyHitObject> difficultyObjects;
+        private readonly Queue<OsuDifficultyHitObject> onScreen = new Queue<OsuDifficultyHitObject>();
+
+        public OsuDifficultyBeatmap(List<OsuHitObject> objects)
+        {
+            // Sort HitObjects by StartTime - they are not correctly ordered in some cases.
+            // This should probably happen before the objects reach the difficulty calculator.
+            objects.Sort((a, b) => a.StartTime.CompareTo(b.StartTime));
+            difficultyObjects = createDifficultyObjectEnumerator(objects);
+        }
+
+        public IEnumerator<OsuDifficultyHitObject> GetEnumerator()

```

Comment:
An xml-doc explaining what the overall goal is here would be great. I understand, that the inner loop goes over all on-screen objects at the moment the hitobject in the outer loop has to be hit, but that would be good to explain and to reason about here.